src/Microsoft.ML.Transforms/Text/NgramTransform.cs

```diff
@@ -607,101 +438,438 @@ private static NgramIdFinder GetNgramIdFinderAdd(int[] counts, int[] lims, Seque
+            Host.CheckValue(ctx, nameof(ctx));
+            ctx.CheckAtModel();
+            ctx.SetVersionInfo(GetVersionInfo());
+            // *** Binary format ***
+            // <base>
+            // for each added column
+            //   _transformInfo
+            //   the ngram SequencePool
+            //   the ngram inverse document frequencies
+            SaveColumns(ctx);
+            for (int i = 0; i < _transformInfos.Length; i++)
+            {
+                _transformInfos[i].Save(ctx);
+                _ngramMaps[i].Save(ctx.Writer);
+                ctx.Writer.WriteDoubleArray(_invDocFreqs[i]);
+            }
         }
 
-        protected override Delegate GetGetterCore(IChannel ch, IRow input, int iinfo, out Action disposer)
+        protected override IRowMapper MakeRowMapper(Schema schema) => new Mapper(this, schema);
+
+        private sealed class Mapper : MapperBase
         {
-            Host.AssertValueOrNull(ch);
-            Host.AssertValue(input);
-            Host.Assert(0 <= iinfo && iinfo < Infos.Length);
-            Host.Assert(Infos[iinfo].TypeSrc.IsVector);
-            Host.Assert(Infos[iinfo].TypeSrc.ItemType.IsKey);
-
-            disposer = null;
-
-            var getSrc = RowCursorUtils.GetVecGetterAs<uint>(NumberType.U4, input, Infos[iinfo].Source);
-            var src = default(VBuffer<uint>);
-            var bldr = new NgramBufferBuilder(_exes[iinfo].NgramLength, _exes[iinfo].SkipLength,
-                _ngramMaps[iinfo].Count, GetNgramIdFinder(iinfo));
-            var keyCount = (uint)Infos[iinfo].TypeSrc.ItemType.KeyCount;
-            if (keyCount == 0)
-                keyCount = uint.MaxValue;
-
-            ValueGetter<VBuffer<Float>> del;
-            switch (_exes[iinfo].Weighting)
+            private readonly ColumnType[] _srcTypes;
+            private readonly int[] _srcCols;
+            private readonly ColumnType[] _types;
+            private readonly NgramTransformer _parent;
+
+            public Mapper(NgramTransformer parent, Schema inputSchema)
+               : base(parent.Host.Register(nameof(Mapper)), parent, inputSchema)
             {
-                case WeightingCriteria.TfIdf:
-                    Host.AssertValue(_invDocFreqs[iinfo]);
-                    del =
-                        (ref VBuffer<Float> dst) =>
-                        {
-                            getSrc(ref src);
-                            if (!bldr.IsEmpty)
+                _parent = parent;
+                _types = new ColumnType[_parent.ColumnPairs.Length];
+                _srcTypes = new ColumnType[_parent.ColumnPairs.Length];
+                _srcCols = new int[_parent.ColumnPairs.Length];
+                for (int i = 0; i < _parent.ColumnPairs.Length; i++)
+                {
+                    _types[i] = new VectorType(NumberType.Float, _parent._ngramMaps[i].Count);
+                    inputSchema.TryGetColumnIndex(_parent.ColumnPairs[i].input, out _srcCols[i]);
```

Comment:
>out _srcCols[i] [](start = 80, length = 15)

this can fail, remember?

The mapper ctor is where you validate schema

Reply:
We do that check in MapperBase

---
In reply to: [233694454](https://github.com/dotnet/machinelearning/pull/1592#discussion_r233694454) [](ancestors = 233694454)